Source/src/modules/ModuleUserInterface.cpp

```diff
@@ -0,0 +1,176 @@
+    App->event->Subscribe(Event::Type::MOUSE_ACTION, handle_mouse_action);
+    return true;
+}
+
+UpdateStatus Hachiko::ModuleUserInterface::Update(float delta)
+{
+    const WindowScene* w_scene = App->editor->GetSceneWindow();
+    ImVec2 mouse_pos = ImGui::GetMousePos();
+    float2 click_pos = w_scene->ImguiToScreenPos(float2(mouse_pos.x, mouse_pos.y));
+    RecursiveCheckMousePos(App->scene_manager->GetActiveScene()->GetRoot(), click_pos, false);
+    return UpdateStatus::UPDATE_CONTINUE;
+}
+
+bool Hachiko::ModuleUserInterface::CleanUp()
+{
+    RemoveSquare();
+    return true;
+}
+
+void Hachiko::ModuleUserInterface::DrawUI(const Scene* scene)
+{
+    Program* program = App->program->GetUserInterfaceProgram();
+    program->Activate();
+    glDepthFunc(GL_ALWAYS);
+    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+    unsigned width, height;
+    App->camera->GetMainCamera()->GetResolution(width, height);
+
+    ModuleProgram::CameraData camera_data;
+    // position data is unused on the ui program
+    camera_data.pos = float3::zero;
+    camera_data.view = float4x4::identity;
+    camera_data.proj = float4x4::D3DOrthoProjLH(-1, 1, static_cast<float>(width), static_cast<float>(height));
+
+    App->program->UpdateCamera(camera_data);
+    BindSquare();
+    RecursiveDrawUI(scene->GetRoot(), program);
+    UnbindSuare();
+    glDepthFunc(GL_LESS);
+}
+
+void Hachiko::ModuleUserInterface::RecursiveDrawUI(GameObject* game_object, Program* program)
+{
+    ComponentCanvasRenderer* renderer = game_object->GetComponent<ComponentCanvasRenderer>();
+
+    if (renderer && game_object->IsActive() && renderer->IsActive())
+    {
+        renderer->Render(program);
+    }
+
+    for (GameObject* child : game_object->children)
+    {
+        RecursiveDrawUI(child, program);
+    }
+}
+
+void Hachiko::ModuleUserInterface::RecursiveCheckMousePos(GameObject* game_object, const float2& mouse_pos, bool is_click)
+{
+    // If it is not click it is considered a hover
+    ComponentTransform2D* transform = game_object->GetComponent<ComponentTransform2D>();
+    // Find selectables TODO: Improve how this is handled
+    ComponentButton* selectable = game_object->GetComponent<ComponentButton>();
+    if (transform && selectable)
+    {
+        if (transform->Intersects(mouse_pos))
+        {
+            //HE_LOG("Intersects %s", transform->GetGameObject()->name.c_str());
+            selectable->OnPointerEnter();
+            if (is_click)
+            {
+                HE_LOG("Clicks %s", transform->GetGameObject()->name.c_str());
+                selectable->Activate();
+            }
+        }
+        else
+        {
+            selectable->OnPointerExit();
+        }
+    }
+    for (GameObject* child : game_object->children)
+    {
+        RecursiveCheckMousePos(child, mouse_pos, is_click);
+    }
+}
+
+void Hachiko::ModuleUserInterface::HandleMouseAction(Hachiko::Event& evt)
```

Comment:
I think it's better to extract the `MouseEventPayload` inside the lambda function and pass it as the parameter to this function instead of passing a generic `Event`